src/main/java/com/lgtm/easymoney/services/impl/LoanServiceImpl.java

```diff
@@ -0,0 +1,190 @@
+    trans.setStatus(TransactionStatus.LOAN_PENDING);
+    return transactionService.saveTransaction(trans);
+  }
+
+  private List<Transaction> getLoansByUser(User user) {
+    List<TransactionStatus> status = List.of(
+        TransactionStatus.LOAN_PENDING,
+        TransactionStatus.LOAN_APPROVED,
+        TransactionStatus.LOAN_DECLINED);
+    return transactionService.getAllTransactionsWithUser(user, status);
+  }
+
+  private Transaction getLoanById(Long id) {
+    return transactionService.getTransactionById(id);
+  }
+
+  private boolean validateLoanRequest(Long lid, Long fromUid, Long toUid) {
+    Transaction t = getLoanById(lid);
+    return t.getFrom().getId().equals(fromUid)
+        && t.getTo().getId().equals(toUid)
+        && t.getStatus() == TransactionStatus.LOAN_PENDING;
+  }
+
+  @Override
+  public ResourceCreatedRsp requestLoan(RequestReq req) {
+    // get params
+    BigDecimal amount = req.getAmount();
+    String category = req.getCategory();
+    String desc = req.getDescription();
+    User borrower = userService.getUserById(req.getFromUid());
+    User lender = userService.getUserById(req.getToUid());
+    // validate if loan user type is valid
+    validateLoanUsers(borrower, lender);
+    // create loan request
+    Transaction loan = createLoanRequest(borrower, lender, amount, desc,
+        Category.valueOf(category.toUpperCase()));
+    // response
+    return loan == null ? new ResourceCreatedRsp(null) : new ResourceCreatedRsp(loan.getId());
+  }
+
+  @Override
+  public LoanRsp getLoansByUid(Long uid) {
+    User user = userService.getUserById(uid);
+    List<Transaction> loans = getLoansByUser(user);
+    // response
+    LoanRsp response = new LoanRsp();
+    response.setSuccess(true);
+    List<TransactionRsp> transferRsps =
+        transactionService.generateListResponseFromTransactions(loans);
+    response.setLoans(transferRsps);
+    response.setMessage("User's loans returned");
+    return response;
+  }
+
+  private boolean approveLoan(Transaction loan) {
+    loan.setStatus(TransactionStatus.TRANS_PENDING);
+    transactionService.saveTransaction(loan);
+    boolean success = transactionService.executeTransaction(loan);
+    if (success) {
+      loan.setStatus(TransactionStatus.LOAN_APPROVED);
+      transactionService.saveTransaction(loan);
+    } else {
+      loan.setStatus(TransactionStatus.LOAN_PENDING);
+      transactionService.saveTransaction(loan);
+    }
+    return success;
+  }
+
+  @Override
+  public LoanRsp approveLoan(RequestAcceptDeclineReq req) {
+    // get params
+    Long fromUid = req.getFromUid();
+    Long toUid = req.getToUid();
+    Long lid = req.getRequestid();
+    // validate
+    User lender = userService.getUserById(fromUid);
+    User borrower = userService.getUserById(toUid);
+    validateLoanUsers(borrower, lender);
+    if (!validateLoanRequest(lid, fromUid, toUid)) {
+      throw new InvalidUpdateException("approve loan", lid, "loanId", lid);
```

Comment:
More precisely, the first arg should just be "Loan" or "Transaction" because it's the resource name, not the operation name

Reply:
Yes lol. Will push the change together with next commit.